pkg/apis/k8sbundle/v1alpha1/bundle_types.go

```diff
@@ -0,0 +1,141 @@
+type ClusterBundleSpec struct {
+	// Version-string for this bundle. The version should be a SemVer string (see
+	// https://semver.org/) of the form X.Y.Z (Major.Minor.Patch).  Generally
+	// speaking, a major-version (changes should indicate breaking changes,
+	// minor-versions should indicate backwards compatible features, and patch
+	// changes should indicate backwords compatible. If there are any changes to
+	// the bundle, then the version string must be incremented.
+	//
+	// If a bundle is versioned, then all its components must be versioned.
+	Version string `json:"version,omitempty"`
+
+	// Kubernetes objects grouped into component packages and versioned together.
+	// These could be applications or they could be some sort of supporting
+	// object collection.
+	Components []ComponentPackage `json:"componentPackage,omitempty"`
+
+	// Cluster components that are specified externally as Files. The process of inlining
+	// for a bundle reads component files into components, and so after
+	// inlining, this list will be empty.
+	ComponentFiles []File `json:"componentFiles,omitempty"`
+}
+
+// File represents some sort of file that's specified external to the bundle,
+// which could be on either a local or remote file system.
+type File struct {
+	// URL to find this file.
+	URL string `json:"url,omitempty"`
+
+	// Optional Sha256 hash of the binary to ensure we are pulling the correct
+	// binary/file.
+	Hash string `json:"hash,omitempty"`
+}
+
+// ComponentPackageSpec represents the spec for the component.
+type ComponentPackageSpec struct {
+	// Version-string for this component. The version should be a SemVer 2 string
+	// (see https://semver.org/) of the form X.Y.Z (Major.Minor.Patch).  A
+	// major-version changes should indicate breaking changes, minor-versions
+	// should indicate backwards compatible features, and patch changes should
+	// indicate backwords compatible. If there are any changes to the component,
+	// then the version string must be incremented.
+	Version string `json:"version,omitempty"`
+
+	// Optional. A version-string representing the version of the API this
+	// component offers to other components (for the purposes of requirement
+	// satisfaction). If no API is offered to other components, then this field
+	// may be blank, but then other components may not depend on this component.
+	ComponentAPIVersion string `json:"componentApiVersion,omitempty"`
+
+	// A list of components that must be packaged in a bundle in with this
+	// component.
+	Requirements []MinRequirement `json:"requirements,omitempty"`
+
+	// Structured Kubenetes objects that run as part of this app, whether on the
+	// master, on the nodes, or in some other fashio.  These Kubernetes objects
+	// are inlined and must be YAML/JSON compatible. Each must have `apiVersion`,
+	// `kind`, and `metadata`.
+	//
+	// This is essentially equivalent to the Kubernetes `Unstructured` type.
```

Comment:
It's RawExtension in API types.

Reply:
If you want to force yaml/json encoding, why not take a list of strings?

Reply:
I've considered it. Then you'd rely on validation, which is certainly possible.

There's a lot of power in having objects forced into a parsed JSON-structure. But it sounds maybe that's not possible in k8S/

Reply:
The thing is, if you don't want people to have to put in quoted yaml/json, you have to do a bunch of tricks during unmarshalling to make it work. quoted json in json is a huge pain, although it's not too bad in yaml.

Anyway, the RawExtension object does this for you.

Reply:
Note to self: https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/types.go#L92

So maybe that's the story.

Reply:
That was one of the beautiful things about using components that are inlined as JSON-format is that it avoids all the quoting nonsense and also is really nice for readability. I'll think about using the RuntimeExtension.